data_management/rocrate.py

```diff
@@ -0,0 +1,661 @@
+"""
+Produce a RO Crate of a CodeRun or DataProduct.
+
+An RO Crate is research object (RO) that has been packaged up, in this case as a zip
+file. This research object is centred around a `CodeRun` or the creation of a
+`DataProduct`.
+
+For a `CodeRun` all output `DataProduct` files are packaged up along with any other
+local files that were used to produce them.
+For the `DataProduct` the `DataProduct` file is packaged up along with any other local
+files that were used to produce it.
+
+Also included in the RO Crate is the metadata file `ro-crate-metadata.json`. The
+`ro-crate-metadata.json` file is made available under the
+[CC0 Public Domain Dedication](https://creativecommons.org/publicdomain/zero/1.0/).
+Please note individual files may have their own licenses.
+All of the packaged files are represented as `File` data entities in the metadata file.
+Any external files will have a link to them in the metadata file, but will not be
+packaged in the zip file.
+
+The `CodeRun` has been modelled as a RO Crate `ContextEntity` of type `CreateAction`,
+see
+[software-used-to-create-files](https://www.researchobject.org/ro-crate/1.1/provenance.html#software-used-to-create-files).
+
+A `CreateAction` has `instrument` property, which represents the software used to
+generate the product. For our purposes `instrument`s include the link to the repo, the
+submission script and configuration file. The submission script metadata is based on
+information from
+[describing-scripts-and-workflows](https://www.researchobject.org/ro-crate/1.1/workflows.html#describing-scripts-and-workflows).
+
+`CreateAction` (`CodeRun`) properties:
+
+* `instrument`: the software used to generate the output
+* `object`: the input files
+* `result`: the output file
+* `agent`: the `Author`
+The RO Crate is available as a `zip` file.
+
+The contents of the ro-crate-metadata file can be viewed as `JSON` or `JSON-LD`.
+
+"""
+from datetime import datetime
+import json
+import mimetypes
+import tempfile
+
+from rocrate.model.person import Person
+from rocrate.rocrate import ContextEntity
+from rocrate.rocrate import ROCrate
+
+from data_management.views import external_object
+
+from . import models
+
+
+RO_TYPE = "@type"
+FILE = "file:"
+
+
+def _add_authors(authors, crate, entity, registry_url):
+    """
+    Add the authors to the crate and associate them with the entity.
+
+    @param authors: a list of authors from the Author table
+    @param crate: the RO Crate object
+    @param entity: the entity to attach the authors to
+    @param registry_url: a str containing the registry URL
+
+    """
+    cr_authors = []
+    for author in authors:
+        author_id = f"{registry_url}api/author/{author.id}"
+        cr_author = crate.add(
+            Person(crate, author_id, properties={"name": author.name})
+        )
+        if author.identifier is not None:
+            cr_author["identifier"] = author.identifier
+        cr_authors.append({"@id": author_id})
```

Comment:
This can simply be `cr_authors.append(cr_author)`, JSON-LD references are handled automatically by `Entity.__setitem__`